File: dns-query-v6/dns_query.py
import dns.resolver
import dns.exception
import json
import concurrent.futures
import csv
import os
from ratelimiter import RateLimiter

def query_domain(device, domain, resolver):
    with resolver['limiter']:
        try:
            # Get the IPv6 address
            # https://dnspython.readthedocs.io/en/latest/resolver-class.html#dns.resolver.Answer
            answer = resolver['resolver'].resolve(domain, 'AAAA') # 'A'
            if answer:
                ipv6_address = ';'.join([rr.to_text() for rr in answer.rrset])
                canonical_name = answer.canonical_name.to_text()
                return (device, domain, ipv6_address, 'no error AAAA response', canonical_name, resolver['resolver'].nameservers[0])
            else:
                if answer.response:
                    return (device, domain, 'No AAAA answer received', 'No AAAA answer received', answer.response.to_text(), resolver['resolver'].nameservers[0])
                else:
                    return (device, domain, 'No AAAA answer received', 'No AAAA answer received', None , resolver['resolver'].nameservers[0])
        except dns.resolver.NoAnswer:
            return (device, domain, 'No AAAA answer received', 'No AAAA answer received', None, resolver['resolver'].nameservers[0])
        except dns.resolver.NXDOMAIN:
            return (device, domain, 'Domain does not exist', 'Domain does not exist', None, resolver['resolver'].nameservers[0])
        except dns.resolver.NoNameservers:
            return (device, domain, 'Nameservers failed to answer', 'Nameservers failed to answer', None, resolver['resolver'].nameservers[0])
        except dns.exception.Timeout:
            return (device, domain, 'Query timed out', 'Query timed out', None, resolver['resolver'].nameservers[0])
        except Exception as e:
            return (device, domain, str(e), 'others', None, resolver['resolver'].nameservers[0])

# Load the JSON file
with open('../dest-analysis/output/all_domain_list_all.json', 'r') as f:
# with open('all_domain_list_all.json', 'r') as f:
    devices = json.load(f)

devices = {k: v for k, v in devices.items()}
print(devices)

# Specify the DNS resolvers
resolvers = [{'resolver': dns.resolver.Resolver(configure=False), 'limiter': RateLimiter(max_calls=5, period=1)} for _ in range(2)]
resolvers[0]['resolver'].nameservers = ['8.8.8.8']  # Google DNS    2001:4860:4860::8888
resolvers[1]['resolver'].nameservers = ['2001:4860:4860::8888']  # Google DNS    2001:4860:4860::8888
# resolvers[0]['resolver'].nameservers = ['208.67.222.222']  # OpenDNS    2620:0:ccc::2
# resolvers[1]['resolver'].nameservers = ['1.1.1.1']  # Cloudflare DNS    2606:4700:4700::1111
# resolvers[3]['resolver'].nameservers = ['223.6.6.6'] # AliDNS   2400:3200::1
# resolvers[4]['resolver'].nameservers = ['155.33.33.75'] # NEU DNS 

# Initialize a dictionary to store the responses
responses = {device: [] for device in devices}

# Prepare the tasks
tasks = [(device, domain, resolver) for device in devices for domain in devices[device] for resolver in resolvers]

# Execute the tasks
with concurrent.futures.ThreadPoolExecutor(max_workers=20) as executor:  # Limit the number of concurrent threads
    future_to_query = {executor.submit(query_domain, device, domain, resolver): (device, domain, resolver) for device, domain, resolver in tasks}
    for future in concurrent.futures.as_completed(future_to_query):
        device, domain, resolver = future_to_query[future]
        response_device, response_domain, response, response_type, canoical_name, nameserver = future.result()
        responses[device].append((response_domain, nameserver, response_type, response, canoical_name))

# Save the responses in CSV files
if not os.path.exists('output'):
    os.makedirs('output')
for device, response_list in responses.items():
    with open(f'output/{device}.csv', 'w', newline='') as f:
        writer = csv.writer(f)
        writer.writerow(['Domain', 'Resolver', 'Response Type', 'Response', 'Canonical Name'])
        for domain, resolver, response_type, response, canoical_name in response_list:
            writer.writerow([domain, resolver, response_type, response, canoical_name])

File: dns-query-v6/dns_query.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"sync"

	"github.com/juju/ratelimit"
	"github.com/miekg/dns"
)

type Device struct {
	Name    string   `json:"name"`
	Domains []string `json:"domains"`
}

type Resolver struct {
	Nameserver string
	Limiter    *ratelimit.Bucket
}

type Response struct {
	Domain        string
	Nameserver    string
	ResponseType  string
	Response      string
	CanonicalName string
}

func queryDomain(device, domain string, resolver Resolver, wg *sync.WaitGroup, ch chan<- Response) {
	defer wg.Done()

	limiter := resolver.Limiter
	limiter.Wait(1)

	client := new(dns.Client)
	msg := new(dns.Msg)
	msg.SetQuestion(dns.Fqdn(domain), dns.TypeAAAA)

	r, _, err := client.Exchange(msg, resolver.Nameserver)
	if err != nil {
		ch <- Response{Domain: domain, Nameserver: resolver.Nameserver, ResponseType: "Query error", Response: err.Error()}
		return
	}

	if len(r.Answer) == 0 {
		ch <- Response{Domain: domain, Nameserver: resolver.Nameserver, ResponseType: "No AAAA answer received", Response: "No AAAA answer received"}
		return
	}

	var ipv6Addresses []string
	var canonicalName string
	for _, ans := range r.Answer {
		switch t := ans.(type) {
		case *dns.AAAA:
			ipv6Addresses = append(ipv6Addresses, t.AAAA.String())
		case *dns.CNAME:
			canonicalName = t.Target
		}
	}

	ch <- Response{Domain: domain, Nameserver: resolver.Nameserver, ResponseType: "no error AAAA response", Response: fmt.Sprintf("%s", ipv6Addresses), CanonicalName: canonicalName}
}

func main() {
	file, err := os.Open("../dest-analysis/output/all_domain_list_all.json")
	if err != nil {
		fmt.Println("Error opening JSON file:", err)
		return
	}
	defer file.Close()

	var devices map[string]Device
	if err := json.NewDecoder(file).Decode(&devices); err != nil {
		fmt.Println("Error decoding JSON:", err)
		return
	}

	// Filter devices
	filteredDevices := make(map[string]Device)
	for k, v := range devices {
		filteredDevices[k] = v
	}
	devices = filteredDevices

	resolvers := []Resolver{
		{Nameserver: "8.8.8.8:53", Limiter: ratelimit.NewBucketWithRate(5, 1)},
		{Nameserver: "[2001:4860:4860::8888]:53", Limiter: ratelimit.NewBucketWithRate(5, 1)},
	}

	var wg sync.WaitGroup
	responseChannel := make(chan Response, len(devices)*len(resolvers))

	for deviceName, device := range devices {
		for _, domain := range device.Domains {
			for _, resolver := range resolvers {
				wg.Add(1)
				go queryDomain(deviceName, domain, resolver, &wg, responseChannel)
			}
		}
	}

	wg.Wait()
	close(responseChannel)

	responses := make(map[string][]Response)
	for response := range responseChannel {
		responses[response.Domain] = append(responses[response.Domain], response)
	}

	for device, responseList := range responses {
		outputFile, err := os.Create(fmt.Sprintf("output/%s.csv", device))
		if err != nil {
			fmt.Println("Error creating CSV file:", err)
			continue
		}
		defer outputFile.Close()

		writer := csv.NewWriter(outputFile)
		defer writer.Flush()

		writer.Write([]string{"Domain", "Resolver", "Response Type", "Response", "Canonical Name"})
		for _, resp := range responseList {
			writer.Write([]string{resp.Domain, resp.Nameserver, resp.ResponseType, resp.Response, resp.CanonicalName})
		}
	}
}